src/vstest.console/Publisher/MetricsPublisher.cs

```diff
@@ -0,0 +1,138 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.VisualStudio.TestPlatform.CommandLine.Publisher
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Globalization;
+
+    using Microsoft.VisualStudio.TestPlatform.ObjectModel;
+
+#if NET451
+    using Microsoft.VisualStudio.Telemetry;
+#endif
+
+    /// <summary>
+    /// The metrics publisher.
+    /// </summary>
+    public class MetricsPublisher : IMetricsPublisher
+    {
+#if NET451
+        private TelemetrySession session;
+#endif
+        public MetricsPublisher()
+        {
+#if NET451
+            try
+            {
+                this.session = new TelemetrySession(TelemetryService.DefaultSession.SerializeSettings());
+                this.session.IsOptedIn = true;
+                this.session.Start();
+            }
+            catch (Exception e)
+            {
+                if (EqtTrace.IsWarningEnabled)
+                {
+                    EqtTrace.Warning(string.Format(CultureInfo.InvariantCulture, "TelemetrySession: Error in starting Telemetry session : {0}", e.Message));
+                }
+            }
+#endif
+        }
+
+            /// <summary>
+            /// Publishes the metrics
+            /// </summary>
+            /// <param name="eventName"></param>
+            /// <param name="metrics"></param>
+            public void PublishMetrics(string eventName, IDictionary<string, string> metrics)
+        {
+#if NET451
+            if (metrics == null || metrics.Count == 0)
+            {
+                return;
+            }
+
+            if (EqtTrace.IsVerboseEnabled)
+            {
+                EqtTrace.Verbose("TelemetrySession: Sending the telemetry data to the server.");
+            }
+
+            try
+            {
+                var finalMetrics = RemoveInvalidCharactersFromProperties(metrics);
+
+                TelemetryEvent telemetryEvent = new TelemetryEvent(eventName);
+
+                foreach (var metric in finalMetrics)
+                {
+                    telemetryEvent.Properties[metric.Key] = metric.Value;
+                }
```

Comment:
If properties is also of same type as your dictionary use it's constructor `new Dictionary<T,T>(Dictionary<T,T> d)`